### PR TITLE
February 2022 IotM: Combat lover's locket

### PR DESCRIPTION
**Pull request by gausie**
This is ready for review if anyone would like to do so!

### CODE REVIEW

File: src/net/sourceforge/kolmafia/session/LocketManager.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+  public static boolean remembersMonster(int monsterId) {
+    return knownMonsters.contains(monsterId);
+  }
+
+  public static void parseMonsters(final String text) {
+    // Visiting the reminisce page is a source of truth
+    knownMonsters.clear();
+
+    var m = REMINISCABLE_MONSTER.matcher(text);
```
Comment (gausie) [inline src/net/sourceforge/kolmafia/session/LocketManager.java:53]: Just a thought, we need to also include monsters you've already fought today as they leave the list.

File: src/net/sourceforge/kolmafia/session/LocketManager.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    foughtMonsters.clear();
+  }
+
+  public static void reset() {
+    if (!own()) {
+      clear();
+      return;
+    }
+
+    RequestThread.postRequest(new GenericRequest("inventory.php?reminisce=1", false));
```
Comment (gausie) [inline src/net/sourceforge/kolmafia/session/LocketManager.java:146]: Does anyone know how I stop this printing "Encounter: <encountername>" to the CLI and session log? I can see how I'd do it, but I'm highly sceptical that there's no other special casing for encounters. I'm sure this is not the first encounter we want to visit silently...

Comment (gausie) [reply src/net/sourceforge/kolmafia/session/LocketManager.java:146]: @Veracity0 do you have any thoughts?

Comment (gausie) [reply src/net/sourceforge/kolmafia/session/LocketManager.java:146]: That looks like it, thanks @midgleyc!

**Review by midgleyc — CHANGES_REQUESTED** [review]

File: src/net/sourceforge/kolmafia/session/LocketManager.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  private static final Pattern REMINISCABLE_MONSTER = Pattern.compile("<option value=\"(\\d+)\"");
+  private static final Set<String> CONSTANT_MODS =
+      Set.of("HP Regen Min", "HP Regen Max", "MP Regen Min", "MP Regen Max", "Single Equip");
+
+  private static void parseFoughtMonsters() {
+    foughtMonsters.clear();
+
+    Arrays.stream(Preferences.getString("_locketMonstersFought").split(","))
+        .filter(Predicate.not(String::isBlank))
+        .map(Integer::parseInt)
```
Comment (Veracity0) [inline src/net/sourceforge/kolmafia/session/LocketManager.java:33]: We have a convention of not assuming that a property has been only modified by KoLmafia itself. What if the user has modified this setting and stuck in cruft?

Perhaps another filter - StringUtilities.isNumeric - say.

Better than throwing a NumberFormatException.

File: src/net/sourceforge/kolmafia/session/LocketManager.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+      // ... we find the indicative modifier in the raw mod string and see what phylum it is
+      // associated with.
+      phylum = locketModString.substring(start, locketModString.indexOf(")", start));
+    }
+
+    Preferences.setString("locketPhylum", phylum);
+  }
+
+  public static boolean own() {
+    return KoLConstants.inventory.contains(ItemPool.get(ItemPool.COMBAT_LOVERS_LOCKET, 1))
```
Comment (Veracity0) [inline src/net/sourceforge/kolmafia/session/LocketManager.java:131]: Don't we have InventoryManager.hasItem for this?

File: src/net/sourceforge/kolmafia/textui/RuntimeLibrary.java [hunk 5]
```diff
@@ -2068,6 +2068,11 @@ public static FunctionList getFunctions() {
             new AggregateType(DataTypes.BOOLEAN_TYPE, DataTypes.MONSTER_TYPE),
             params));
 
+    params = new Type[] {};
+    functions.add(
+        new LibraryFunction(
+            "get_locket_monsters", new AggregateType(DataTypes.MONSTER_TYPE, 0), params));
```
Comment (Veracity0) [inline src/net/sourceforge/kolmafia/textui/RuntimeLibrary.java:2074]: Isn't your method returning a map from monster => boolean?

File: test/net/sourceforge/kolmafia/request/FightRequestTest.java [hunk 6]
```diff
@@ -206,17 +205,37 @@ public void cursedMagnifyingGlassTest() throws IOException {
...
   public void daylightShavingTest() throws IOException {
-    KoLCharacter.reset("the Tristero");
     EquipmentManager.setEquipment(
         EquipmentManager.HAT, ItemPool.get(ItemPool.DAYLIGHT_SHAVINGS_HELMET));
     parseCombatData("request/test_fight_daylight_shavings_buff.html");
     assertEquals(2671, Preferences.getInteger("lastBeardBuff"));
   }
 
   @Test
-  public void luckyGoldRingVolcoinoDropRecorded() throws IOException {
```
[the remark this answers is not in the document]
Comment (Veracity0) [reply test/net/sourceforge/kolmafia/request/FightRequestTest.java:217]: I'll be a little more forceful.
Do not remove existing tests.

**Review by Veracity0 — CHANGES_REQUESTED** [review]

File: test/net/sourceforge/kolmafia/request/FightRequestTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (gausie) [reply test/net/sourceforge/kolmafia/request/FightRequestTest.java:217]: That was not intended, I have no idea how that happened honestly

File: src/net/sourceforge/kolmafia/session/LocketManager.java (hunk 4 above)
Comment (gausie) [reply src/net/sourceforge/kolmafia/session/LocketManager.java]: FIxed this in a few other places too

**Pull request merged**